src/decode.rs

```diff
@@ -3682,1522 +3693,119 @@ unsafe fn decode_b(
-                    as *mut alias32))
-                    .u32_0 = 0 as libc::c_int as uint32_t;
-                (*(&mut *(t.l.intra).as_mut_ptr().offset(by4 as isize) as *mut uint8_t
-                    as *mut alias32))
-                    .u32_0 = 0x1010101 as libc::c_uint;
-                (*(&mut *(t.l.skip).as_mut_ptr().offset(by4 as isize) as *mut uint8_t
-                    as *mut alias32))
-                    .u32_0 = (0x1010101 as libc::c_uint).wrapping_mul(b.skip as libc::c_uint);
-                (*(&mut *(*(t.pal_sz_uv).as_mut_ptr().offset(1))
-                    .as_mut_ptr()
-                    .offset(by4 as isize) as *mut uint8_t as *mut alias32))
-                    .u32_0 = (0x1010101 as libc::c_uint).wrapping_mul(
-                    (if has_chroma {
-                        b.c2rust_unnamed.c2rust_unnamed.pal_sz[1] as libc::c_int
-                    } else {
-                        0 as libc::c_int
-                    }) as libc::c_uint,
+
+        // update contexts
+
+        let y_mode_nofilt = if b.y_mode() == FILTER_PRED as u8 {
+            DC_PRED as IntraPredMode
+        } else {
+            b.y_mode() as IntraPredMode
+        };
+
+        let mut set_ctx = |dir: &mut BlockContext, diridx, off, mul, rep_macro: SetCtxFn| {
+            // NOTE: This corresponds to the following logic in the original C:
+            //
+            // ```
+            // ((uint8_t *) &t_dim->lw)[diridx]
+            // ```
+            //
+            // The original logic casts a pointer to the `lw` field to a `*const u8` and
+            // then does indexing within the struct's fields. This is deeply cursed and
+            // unnecessary since `diridx` will only ever be 0 or 1. So for the Rust version
+            // we manually select which field we want based on `diridx` directly.
+            let lw_lh = match diridx {
+                0 => t_dim.lw,
+                1 => t_dim.lh,
+                _ => unreachable!(),
+            };
+
+            rep_macro(
+                dir.tx_intra.as_mut_ptr() as *mut u8,
+                off,
+                mul * lw_lh as u64,
+            );
+            rep_macro(dir.tx.as_mut_ptr() as *mut u8, off, mul * lw_lh as u64);
+            rep_macro(dir.mode.as_mut_ptr(), off, mul * y_mode_nofilt as u64);
+            rep_macro(dir.pal_sz.as_mut_ptr(), off, mul * b.pal_sz()[0] as u64);
+            rep_macro(dir.seg_pred.as_mut_ptr(), off, mul * seg_pred as u64);
+            rep_macro(dir.skip_mode.as_mut_ptr(), off, 0);
+            rep_macro(dir.intra.as_mut_ptr(), off, mul);
+            rep_macro(dir.skip.as_mut_ptr(), off, mul * b.skip as u64);
+
+            // see aomedia bug 2183 for why we use luma coordinates here
+            rep_macro(
+                t.pal_sz_uv[diridx].as_mut_ptr(),
+                off,
+                mul * if has_chroma { b.pal_sz()[1] as u64 } else { 0 },
+            );
+
+            if is_inter_or_switch(frame_hdr) {
```

Comment:
```suggestion
            rep_macro(dir.skip.as_mut_ptr(), off, mul * b.skip as u64);
            // see aomedia bug 2183 for why we use luma coordinates here
            rep_macro(
                t.pal_sz_uv[diridx].as_mut_ptr(),
                off,
                mul * if has_chroma { b.pal_sz()[1] as u64 } else { 0 },
            );
            if is_inter_or_switch(frame_hdr) {
```

I've been trying to keep the extra line breaks the same as the C, as they may convey some semantic meaning.